radix-engine-stores/src/interface.rs

```diff
@@ -0,0 +1,259 @@
+/*
+
+High-level Abstraction
+
++-------------------------+
+|       Radix Engine      |
+|----> SubstateStore <----|
+|          Track          |
+|---> SubstateDatabase <--|
+|         Database        |
++-------------------------+
+
+*/
+
+use radix_engine_interface::api::LockFlags;
+use radix_engine_interface::crypto::*;
+use radix_engine_interface::types::*;
+use radix_engine_interface::*;
+use sbor::rust::prelude::*;
+
+// TODO: Add streaming support for `list_substates`
+
+/// Utility function for encoding a substate ID `(NodeId, ModuleId, SubstateKey)` into a `Vec<u8>`,
+pub fn encode_substate_id(
+    node_id: &NodeId,
+    module_id: ModuleId,
+    substate_key: &SubstateKey,
+) -> Vec<u8> {
+    let mut buffer = Vec::new();
+    buffer.extend(node_id.as_ref());
+    buffer.push(module_id.0);
+    buffer.extend(substate_key.as_ref()); // Length is marked by EOF
+    buffer
+}
+
+/// Utility function for decoding a substate ID `(NodeId, ModuleId, SubstateKey)` from a `Vec<u8>`,
+pub fn decode_substate_id(slice: &[u8]) -> Option<(NodeId, ModuleId, SubstateKey)> {
+    if slice.len() >= NodeId::LENGTH + 1 {
+        // Decode node id
+        let mut node_id = [0u8; NodeId::LENGTH];
+        node_id.copy_from_slice(&slice[0..NodeId::LENGTH]);
+        let node_id = NodeId(node_id);
+
+        // Decode module id
+        let module_id = ModuleId(slice[NodeId::LENGTH]);
+
+        // Decode substate key
+        if let Some(substate_key) = SubstateKey::from_slice(&slice[NodeId::LENGTH + 1..]) {
+            return Some((node_id, module_id, substate_key));
+        }
+    }
+    return None;
+}
+
+/// Error when acquiring a lock.
+#[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor)]
+pub enum AcquireLockError {
+    NotFound(NodeId, ModuleId, SubstateKey),
+    SubstateLocked(NodeId, ModuleId, SubstateKey),
+    LockUnmodifiedBaseOnNewSubstate(NodeId, ModuleId, SubstateKey),
+    LockUnmodifiedBaseOnOnUpdatedSubstate(NodeId, ModuleId, SubstateKey),
+}
+
+/// Represents the interface between Radix Engine and Track.
+///
+/// In practice, we will likely end up with only one implementation.
+///
+/// The trait here is for formalizing the interface and intended user flow.
+pub trait SubstateStore {
+    /// Acquires a lock over a substate.
+    ///
+    /// # Panics
+    /// - If the module ID is invalid
+    fn acquire_lock(
```

Comment:
I'm wondering if we even need locks at this layer. It was initially implemented as a sanity check.